ledger_device_sdk/src/nbgl.rs

```diff
@@ -473,6 +495,570 @@ impl<'a, const MAX_FIELD_NUMBER: usize, const STRING_BUFFER_SIZE: usize>
+}
+
+/// Structure exposed by the NBGL Rust API to the user to create a
+/// button to confirm some information that will be displayed
+/// on the device when using the NbglGenericReview struct.
+pub struct InfoButton<'a> {
+    pub text: &'a str,
+    pub icon: Option<&'a NbglGlyph<'a>>,
+    pub button_text: &'a str,
+    pub tune_id: TuneIndex,
+}
+
+/// Wrapper around the nbgl_contentInfoButton_t struct
+/// that contains the necessary variables used as pointers
+/// in the C struct.
+/// The variables are heap allocated and stored
+/// in the wrapper to ensure they live long enough.
+#[allow(dead_code)]
+struct InfoButtonWrapper {
+    c_struct: nbgl_contentInfoButton_t,
+    text: CString,
+    button_text: CString,
+    icon: Option<Box<nbgl_icon_details_t>>,
+    content_type: nbgl_contentType_t,
+    action_callback: nbgl_contentActionCallback_t,
+}
+
+impl InfoButtonWrapper {
+    // Get a reference to the internal nbgl_contentInfoButton_t struct
+    pub fn get_c_struct_ref(&self) -> &nbgl_contentInfoButton_t {
+        &self.c_struct
+    }
+}
+
+impl<'a> From<InfoButton<'a>> for InfoButtonWrapper {
+    fn from(info: InfoButton) -> Self {
+        // Create heap allocated variables that will live long
+        // enough in the wrapper to be used in the C struct when
+        // passed to the C API.
+        let text = CString::new(info.text).unwrap();
+        let button_text = CString::new(info.button_text).unwrap();
+        let mut icon_ptr: *const nbgl_icon_details_t = core::ptr::null();
+        let mut icon_boxed: Option<Box<nbgl_icon_details_t>> = None;
+        if info.icon.is_some() {
+            icon_boxed = Some(Box::new(info.icon.unwrap().into()));
+            icon_ptr = icon_boxed.as_ref().unwrap().as_ref();
+        };
+        InfoButtonWrapper {
+            c_struct: nbgl_contentInfoButton_t {
+                text: text.as_ptr() as *const c_char,
+                icon: icon_ptr,
+                buttonText: button_text.as_ptr() as *const c_char,
+                buttonToken: FIRST_USER_TOKEN as u8,
+                tuneId: info.tune_id as u8,
+            },
+            text: text,
+            button_text: button_text,
+            icon: icon_boxed,
+            content_type: INFO_BUTTON,
+            action_callback: Some(generic_content_action_callback),
+        }
+    }
+}
+
+/// Structure exposed by the NBGL Rust API to the user to create a
+/// tag/value list screen that will be displayed on the device when
+/// using the NbglGenericReview struct.
+pub struct TagValueList<'a> {
+    pub pairs: &'a [Field<'a>],
+    pub nb_max_lines_for_value: u8,
+    pub small_case_for_value: bool,
+    pub wrapping: bool,
+}
+
+/// Wrapper around the nbgl_contentTagValueList_t struct
+/// that contains the necessary variables used as pointers
+/// in the C struct.
+/// The variables are heap allocated and stored
+/// in the wrapper to ensure they live long enough.
+#[allow(dead_code)]
+struct TagValueListWrapper {
```

Comment:
I do not understand why have you defined all those wrappers ? And why storing all the variables in the heap ?

Reply:
Because the c_struct takes pointers to these variables... If I don't store them in the wrapper (as CString for example, which is why I mean by "on the heap"), then they don't live long enough... And the app crashes (or displays weird stuff)

Reply:
@yogh333 either we have wrappers to store variables or the API needs to ask for references (such as `&CStr`) to these variables. But we decided we would allow the user to use `&str` and not bother with `CString` and such.

Reply:
Using heap is not mandatory I think. For instance, take a look on my [branch](https://github.com/LedgerHQ/ledger-device-rust-sdk/blob/568e3d88ee8326dee7992f0b6e79968b95910dec/ledger_device_sdk/src/nbgl.rs#L351C18-L351C32): `tag_value_list` is defined on the stack just before calling the NBGL C function `ux_sync_review`. As long as the scope of `tag_value_list` is not over before calling the C function, it is okay. Same for `Cstring`actually: if the CString lives on the stack and is in the same scope as the C NBGL function call, then there is no pb to pass the CString as a pointer (`CString.as_ptr()`): see for instance [here](https://github.com/LedgerHQ/ledger-device-rust-sdk/blob/568e3d88ee8326dee7992f0b6e79968b95910dec/ledger_device_sdk/src/nbgl.rs#L409).

Reply:
@yogh333 of course it works in the simple case of a CString (which by the way is allocated on the heap anyway even if you declare it locally in a function, like String), because it lives long enough in the function `show` : 

* function show is called,
* CString is allocated,
* C Api function is called with pointer to CString,
* C function returns
* show function returns, CString is dropped

In my case the sequence is a lot different : 

* `addcontent` of the generic review is called with a content struct such as  `TagValueList` as parameter. The `addcontent` function takes ownership of the `TagValueList` variable and converts it to the wrapper structure with the `impl<'a> From<TagValueList<'a>> for TagValueListWrapper` trait function.
* Variables are created (such as `let mut c_field_names: Vec<CString> = Vec::new();`) to store the different field names to be displayed. We actually don't care they are on the heap, the important thing is they are stored in the wrapper.
* Pointers to these variables are used to intialize the fields of the underlying C struct.
* The `From` trait function **returns** the wrapper which contains the C struct and the actual variables it points to, ensuring they live long enough. The original `TagValueList` (for instance) is dropped.
* The wrapper is added to the Vec in the NbglGenericReview struct, which can then use this list to display the different screens when the show function is called.

We are not in the case of a single `show` function called, we need `add_content` + `From` trait function to perform conversions

If you want to discuss a better implementation I'd very much appreciate if we do it *before*  we merge the PR. I wouldn't like if the next day you rewrite my code because it doesn't suit you.

Reply:
@yogh333 If we use a trait function to convert from let's say `TagValueList` to a `nbgl_contentTagValueList_t` (the C struct) directly, we can declare variables and use the pointers to these variables in the `nbgl_contentTagValueList_t`, but once the conversion function returns, these variables have nowhere to go and will be dropped...

I tried.

Reply:
I will update the code comments to avoid mentioning `heap` everywhere. We actually don't care it's on the heap. The important thing is we need to ensure these variables live long enough to be used when we actually call the C api.

Reply:
@agrojean-ledger Thanks for your explanation. However, we must challenge the need to store **all** the contents before calling `show`. Why not calling `show` for each content, and thus giving up the `add_content` function ? In that case, a conversion from `TagValueList` to `nbgl_contentTagValueList_t ` in the show function is enough.

Reply:
@yogh333 the C generic review API works by passing a `contentsList: *const nbgl_content_t,`, which has to contain all the different content structures... 

There's a possibility to avoid having to preallocate everything by using the `contentGetterCallback`, when the app does not have knowledge of all the contents before calling the function. 

The API uses a union to allow both behaviors : 

```
pub union nbgl_genericContents_t__bindgen_ty_1 {
    pub contentsList: *const nbgl_content_t,
    pub contentGetterCallback: nbgl_contentCallback_t,
}
```

But as its name implies, the second option uses a *callback*. I didn't put too much effort looking at this one because we know : 

1. Using callbacks is a pain (may not even be feasible with the current C ux_sync library),
2. It's another kind of usage (where the app does not have all the content ready before displaying the review)

If you want a "typical" example of generic review usage, take a look at the ethereum app : https://github.com/LedgerHQ/app-ethereum/blob/76e8690a68fb84910d09fbfd14cc975cefabe41a/src_nbgl/ui_display_privacy.c#L23
You can see, everything is allocated before calling the generic review API.

Reply:
@yogh333 I will try with the callback.

Reply:
Sorry but I think you miss my point. After looking into the C API, [nbgl_useCaseGenericReview](https://github.com/LedgerHQ/ledger-secure-sdk/blob/1aa3c0a56c748dc6085afdee2d489a66257d5ee7/lib_nbgl/src/nbgl_use_case.c#L2621) uses `nbgl_genericContents_t`:
```
typedef struct {
    bool callbackCallNeeded;  ///< indicates whether contents should be retrieved using
                              ///< contentsList or contentGetterCallback
    union {
        const nbgl_content_t *contentsList;  ///< array of nbgl_content_t (nbContents items).
        nbgl_contentCallback_t
            contentGetterCallback;  ///< function to call to retrieve a given content
    };
    uint8_t nbContents;  ///< number of contents
} nbgl_genericContents_t;
```
In this struct, we can use `nbgl_content_t` which needs `nbgl_contentTagValueList_t`. So I think we can intialize all those C structs just before calling the C function `nbgl_useCaseGenericReview`, i.e when calling the Rust `show` function. To me there is no need to use a  `TagValueListWrapper`. Let's assume we keep the `add_content` function: `add_content` function could store all the contents using Rust types (`String`, `&str`) but the final conversion is directly performed when calling the `show` function, right before calling the `nbgl_useCaseGenericReview`. One might also think about the Rust `show` function that passes **all** the content (using Rust types/structures) to be displayed, and once again all the conversion are performed just before calling `nbgl_useCaseGenericReview`. Hope it clarifies my proposal 🙏  (and yes callback is a pain, hope I have managed to save you from that hell 😉 )